From 7e4b3657b07e25d352d7c1d0d1dcb5100df0469b Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sun, 29 Oct 2017 18:37:19 -0400
Subject: [PATCH] Add system property to disable book size limits

If anyone comes in with a watchdog crash related to books after this patch
you will not only be publicly shamed but also made an example of.

Disables the security limits on books entirely, allowing plugins AND players
to make books with as much data as they want. Do not use this without
limiting incoming data from packets in some other way.

diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
index b047b4b..fc8439f 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
@@ -32,6 +32,7 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
     static final int MAX_PAGES = 50;
     static final int MAX_PAGE_LENGTH = 320;
     static final int MAX_TITLE_LENGTH = 32;
+    private static final boolean OVERRIDE_CHECKS = Boolean.getBoolean("disable.book-limits"); // Paper - Add override
     protected String title;
     protected String author;
     public List<ITextComponent> pages;
@@ -361,7 +362,7 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
         if (title == null) {
             this.title = null;
             return true;
-        } else if (title.length() > 32) {
+        } else if (title.length() > MAX_TITLE_LENGTH && !OVERRIDE_CHECKS) { // Paper - Add override, use constant for title length
             return false;
         } else {
             this.title = title;
@@ -394,7 +395,7 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
         if (!this.isValidPage(page)) {
             throw new IllegalArgumentException("Invalid page number " + page + "/" + this.pages.size());
         } else {
-            String newText = text == null ? "" : (text.length() > 320 ? text.substring(0, 320) : text);
+            String newText = text == null ? "" : (text.length() > MAX_PAGE_LENGTH && !OVERRIDE_CHECKS ? text.substring(0, MAX_PAGE_LENGTH) : text);
             this.pages.set(page - 1, CraftChatMessage.fromString(newText, true)[0]);
         }
     }
@@ -406,14 +407,14 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
 
     public void addPage(String... pages) {
         for (String page : pages) {
-            if (this.pages.size() >= 50) {
+            if (this.pages.size() >= MAX_PAGES && !OVERRIDE_CHECKS) { // Paper - Add override, use constant for max pages
                 return;
             }
 
             if (page == null) {
                 page = "";
-            } else if (page.length() > 320) {
-                page = page.substring(0, 320);
+            } else if (page.length() > MAX_PAGE_LENGTH && !OVERRIDE_CHECKS) { // Paper - Add override, use constant for page length
+                page = page.substring(0, MAX_PAGE_LENGTH); // Paper - Use constant for page length
             }
 
             this.pages.add(CraftChatMessage.fromString(page, true)[0]);
-- 
2.14.3

